tower-http/src/timeout/body.rs

```diff
@@ -0,0 +1,369 @@
+//! Middleware that applies a timeout to bodies.
+//!
+//! Bodies must produce data at most within the specified timeout.
+//! If they are inactive, an error will be generated.
+//!
+//! # Differences from `tower_http::timeout::service::Timeout`
+//!
+//! `tower_http::timeout::service::Timeout` applies a timeout on the full request.
+//! That timeout is not reset when bytes are handled, whether the request is active or not.
+//!
+//! This middleware will return a [`TimeoutError`].
+//!
+//! # Example
+//!
+//! ```
+//! use http::{Request, Response};
+//! use hyper::Body;
+//! use std::time::Duration;
+//! use tower::ServiceBuilder;
+//! use tower_http::timeout::body::RequestBodyTimeoutLayer;
+//!
+//! async fn handle(_: Request<Body>) -> Result<Response<Body>, std::convert::Infallible> {
+//!     // ...
+//!     # todo!()
+//! }
+//!
+//! # #[tokio::main]
+//! # async fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! let svc = ServiceBuilder::new()
+//!     // Timeout bodies after 30 seconds of inactivity
+//!     .layer(RequestBodyTimeoutLayer::new(Duration::from_secs(30)))
+//!     .service_fn(handle);
+//! # Ok(())
+//! # }
+//! ```
+
+use crate::BoxError;
+use futures_core::{ready, Future};
+use http::{Request, Response};
+use http_body::Body;
+use pin_project_lite::pin_project;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio::time::{sleep, Sleep};
+use tower_layer::Layer;
+use tower_service::Service;
+
+pin_project! {
+    /// Wrapper around a [`http_body::Body`] to time out if data is not ready within the specified duration.
+    pub struct TimeoutBody<B> {
+        timeout: Duration,
+        #[pin]
+        sleep: Option<Sleep>,
```

Comment:
So I think there is a small issue here in that body 0.4 doesn't define what happens if when poll_data returns Pending that if you should be able to poll_trailers. So in that case, this single timeout could fire early in some awkward situations where someone is polling both data and trailers. So the solution here would be to have two sleep states in the body wrapper. I think we need to do this now for the current release but I think we should also add a note about how in `http-body` 1.0 these two poll_fn will be merged into a single `poll_frame` and that we can go back to a single sleep impl. So a comment about that would be helpful.

Reply:
Added, thanks!